### ISSUE TITLE
OutOfMemoryError when scrolling through list of large images on Android L

### ISSUE THREAD

**Issue by tiftof** [MvvmCross/MvvmCross#848]
I'm using MvxImageView to show images in a MvxListView on Android. Images are 800*600 in size. It's not unusual to have a couple of hundreds of those images (all unique) in one list. Everything works fine on pre Lollipop devices. Sometimes there is a delay between the item being revealed in the list and the image being shown, but that is to be expected when scrolling through such a large list of images as not all of them can be contained in the cache.

However, on Lollipop I'm experiencing crashes when scrolling back and forth inside the list. Crashes occur when I scroll over a range of images that is too big to be contained in the cache => images need to be refetched. The crash log is appended at the end of this description. The java heap seems to run out of memory when allocating bitmaps. When using the Android Debug Monitor to monitor the heap size, I can see the heap growing steadily whenever a cache miss occurs and an image needs to be refetched. The allocation tracker reveals that new memory is constantly being allocated by android.graphics.BitmapFactory.decodeByteArray() to reload the images. But it seems like the memory of previously loaded images, but already removed from cache, is not being freed.

While investigating this problem I came across [this interesting line of code](https://github.com/MvvmCross/MvvmCross/blob/bbf9a2ac76e74d9404f4b57036c6e29dfe2cc6c3/Plugins/Cirrious/DownloadCache/Cirrious.MvvmCross.Plugins.DownloadCache.Droid/MvxAndroidLocalFileImageLoader.cs#L72). According to [this blog post](http://slodge.blogspot.co.uk/2013/02/huge-android-memory-bug-and-bug-hunting.html), this is an important line of code to prevent Android filling up memory when loading bitmaps. However, the Android API documentation mentions that this inPurgeable option is [deprecated and ignored in Lollipop](http://developer.android.com/reference/android/graphics/BitmapFactory.Options.html#inPurgeable).

It seems like this introduces a big memory leak again in Lollipop. (Although I don't fully understand yet why this option is needed, I would expect the bitmap memory to be freed when no reference to it is being kept from within the cache anymore)

Log:

```
[Mono] GC_TAR_BRIDGE bridges 157 objects 157 colors 157 ignored 0 sccs 157 xref 0 cache 0/0 setup 0.18ms tarjan 0.15ms scc-setup 0.18ms gather-xref 0.00ms xref-setup 0.00ms cleanup 0.09ms
[Mono] GC_MINOR: (Nursery full) pause 11.57ms, total 12.12ms, bridge 2.69ms promoted 160K major 2048K los 5522K
[art] Background partial concurrent mark sweep GC freed 372(22KB) AllocSpace objects, 1(58KB) LOS objects, 7% free, 195MB/211MB, paused 5.493ms total 23.773ms
[Mono] GC_TAR_BRIDGE bridges 206 objects 206 colors 206 ignored 0 sccs 206 xref 0 cache 0/0 setup 0.21ms tarjan 0.21ms scc-setup 0.15ms gather-xref 0.03ms xref-setup 0.00ms cleanup 0.06ms
[Mono] GC_MINOR: (Nursery full) pause 13.61ms, total 13.76ms, bridge 5.00ms promoted 48K major 2096K los 9483K
[art] Alloc sticky concurrent mark sweep GC freed 679(28KB) AllocSpace objects, 8(440KB) LOS objects, 0% free, 510MB/511MB, paused 1.159ms total 5.554ms
[art] Clamp target GC heap from 526MB to 512MB
[art] Alloc partial concurrent mark sweep GC freed 371(20KB) AllocSpace objects, 1(104KB) LOS objects, 0% free, 510MB/512MB, paused 854us total 13.092ms
[art] Clamp target GC heap from 527MB to 512MB
[art] WaitForGcToComplete blocked for 12.420ms for cause Alloc
[art] Clamp target GC heap from 527MB to 512MB
[art] Alloc partial concurrent mark sweep GC freed 83(3KB) AllocSpace objects, 1(33KB) LOS objects, 0% free, 511MB/512MB, paused 885us total 10.955ms
[art] Clamp target GC heap from 527MB to 512MB
[art] Alloc concurrent mark sweep GC freed 153(6KB) AllocSpace objects, 0(0B) LOS objects, 0% free, 511MB/512MB, paused 823us total 19.958ms
[art] Forcing collection of SoftReferences for 1875KB allocation
[art] Clamp target GC heap from 527MB to 512MB
[art] Alloc concurrent mark sweep GC freed 126(6KB) AllocSpace objects, 0(0B) LOS objects, 0% free, 511MB/512MB, paused 793us total 19.042ms
[art] WaitForGcToComplete blocked for 49.011ms for cause Background
[art] Throwing OutOfMemoryError "Failed to allocate a 1920012 byte allocation with 72503 free bytes and 70KB until OOM"
[skia] --- decoder->decode returned false
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Alloc partial concurrent mark sweep GC freed 28(1216B) AllocSpace objects, 0(0B) LOS objects, 0% free, 511MB/512MB, paused 915us total 10.833ms
[art] Alloc sticky concurrent mark sweep GC freed 0(0B) AllocSpace objects, 0(0B) LOS objects, 0% free, 511MB/512MB, paused 1.342ms total 7.965ms
[art] WaitForGcToComplete blocked for 17.059ms for cause Background
[art] Clamp target GC heap from 527MB to 512MB
mvx:Diagnostic: 93.63 failed to download image http://x.x.com/x.jpg : OutOfMemoryError: Exception of type 'Java.Lang.OutOfMemoryError' was thrown.
      at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw () [0x00000] in <filename unknown>:0 
  at Android.Runtime.JNIEnv.CallStaticObjectMethod (IntPtr jclass, IntPtr jmethod, Android.Runtime.JValue[] parms) [0x00000] in <filename unknown>:0 
  at Android.Graphics.BitmapFactory.DecodeByteArray (System.Byte[] data, Int32 offset, Int32 length, Android.Graphics.Options opts) [0x00000] in <filename unknown>:0 
  at Cirrious.MvvmCross.Plugins.DownloadCache.Droid.MvxAndroidLocalFileImageLoader.LoadBitmap (System.String localPath) [0x00000] in <filename unknown>:0 
  at Cirrious.MvvmCross.Plugins.DownloadCache.Droid.MvxAndroidLocalFileImageLoader.Load (System.String localPath, Boolean shouldCache) [0x00000] in <filename unknown>:0 
  at Cirrious.MvvmCross.Plugins.DownloadCache.MvxImageCache`1[Android.Graphics.Bitmap].Parse (System.String path) [0x00000] in <filename unknown>:0 
  at Cirrious.MvvmCross.Plugins.DownloadCache.MvxImageCache`1[Android.Graphics.Bitmap].ProcessFilePath (System.String url, System.String filePath) [0x00000] in <filename unknown>:0 
  --- End of managed exception stack trace ---
java.lang.OutOfMemoryError: Failed to allocate a 1920012 byte allocation with 72503 free bytes and 70KB until OOM
    at dalvik.system.VMRuntime.newNonMovableArray(Native Method)
    at android.graphics.BitmapFactory.nativeDecodeByteArray(Native Method)
    at android.graphics.BitmapFactory.decodeByteArray(BitmapFactory.java:522)
[art] Clamp target GC heap from 527MB to 512MB
[art] WaitForGcToComplete blocked for 21.667ms for cause Alloc
[art] Clamp target GC heap from 527MB to 512MB
[art] Alloc concurrent mark sweep GC freed 63(2592B) AllocSpace objects, 0(0B) LOS objects, 0% free, 511MB/512MB, paused 1.922ms total 20.324ms
[art] Forcing collection of SoftReferences for 1875KB allocation
[art] WaitForGcToComplete blocked for 12.695ms for cause Background
[art] Clamp target GC heap from 527MB to 512MB
[art] WaitForGcToComplete blocked for 14.770ms for cause Alloc
[art] Clamp target GC heap from 527MB to 512MB
[art] Alloc concurrent mark sweep GC freed 6(288B) AllocSpace objects, 0(0B) LOS objects, 0% free, 511MB/512MB, paused 579us total 18.951ms
[art] WaitForGcToComplete blocked for 17.944ms for cause Background
[art] Throwing OutOfMemoryError "Failed to allocate a 1920012 byte allocation with 19736 free bytes and 19KB until OOM"
[skia] --- decoder->decode returned false
[art] Clamp target GC heap from 527MB to 512MB
[art] WaitForGcToComplete blocked for 8.544ms for cause Alloc
[art] Clamp target GC heap from 527MB to 512MB
[art] Alloc partial concurrent mark sweep GC freed 10(288B) AllocSpace objects, 0(0B) LOS objects, 0% free, 511MB/512MB, paused 1.037ms total 10.955ms
[art] WaitForGcToComplete blocked for 14.465ms for cause Background
[art] Clamp target GC heap from 527MB to 512MB
[art] Alloc concurrent mark sweep GC freed 7(224B) AllocSpace objects, 0(0B) LOS objects, 0% free, 511MB/512MB, paused 3.906ms total 26.275ms
[art] WaitForGcToComplete blocked for 16.601ms for cause Background
[art] Forcing collection of SoftReferences for 52KB allocation
[art] Clamp target GC heap from 527MB to 512MB
[art] WaitForGcToComplete blocked for 22.155ms for cause Alloc
[art] Clamp target GC heap from 527MB to 512MB
[art] Alloc concurrent mark sweep GC freed 6(192B) AllocSpace objects, 0(0B) LOS objects, 0% free, 511MB/512MB, paused 640us total 19.531ms
[art] WaitForGcToComplete blocked for 13.244ms for cause Background
[art] Throwing OutOfMemoryError "Failed to allocate a 53302 byte allocation with 19352 free bytes and 18KB until OOM"
[art] art/runtime/check_jni.cc:65] JNI DETECTED ERROR IN APPLICATION: JNI SetByteArrayRegion called with pending exception 'java.lang.OutOfMemoryError' thrown in unknown throw location
[art] art/runtime/check_jni.cc:65]     in call to SetByteArrayRegion
[art] art/runtime/check_jni.cc:65] "Thread-767" prio=5 tid=22 Runnable
[art] art/runtime/check_jni.cc:65]   | group="main" sCount=0 dsCount=0 obj=0x12e0f080 self=0xafc36800
[art] art/runtime/check_jni.cc:65]   | sysTid=16692 nice=0 cgrp=apps sched=0/0 handle=0x9f90b300
[art] art/runtime/check_jni.cc:65]   | state=R schedstat=( 0 0 0 ) utm=93 stm=31 core=0 HZ=100
[art] art/runtime/check_jni.cc:65]   | stack=0x9e4e1000-0x9e4e3000 stackSize=1020KB
[art] art/runtime/check_jni.cc:65]   | held mutexes= "mutator lock"(shared held)
[art] art/runtime/check_jni.cc:65]   native: #00 pc 00004c58  /system/lib/libbacktrace_libc++.so (UnwindCurrent::Unwind(unsigned int, ucontext*)+23)
[art] art/runtime/check_jni.cc:65]   native: #01 pc 000034c1  /system/lib/libbacktrace_libc++.so (Backtrace::Unwind(unsigned int, ucontext*)+8)
[art] art/runtime/check_jni.cc:65]   native: #02 pc 0025918d  /system/lib/libart.so (art::DumpNativeStack(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, int, char const*, art::mirror::ArtMethod*)+84)
[art] art/runtime/check_jni.cc:65]   native: #03 pc 0023cd13  /system/lib/libart.so (art::Thread::Dump(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) const+162)
[art] art/runtime/check_jni.cc:65]   native: #04 pc 000b1195  /system/lib/libart.so (art::JniAbort(char const*, char const*)+620)
[art] art/runtime/check_jni.cc:65]   native: #05 pc 000b18c5  /system/lib/libart.so (art::JniAbortF(char const*, char const*, ...)+68)
[art] art/runtime/check_jni.cc:65]   native: #06 pc 000b4b4d  /system/lib/libart.so (art::ScopedCheck::ScopedCheck(_JNIEnv*, int, char const*)+1324)
[art] art/runtime/check_jni.cc:65]   native: #07 pc 000be81b  /system/lib/libart.so (art::CheckJNI::SetByteArrayRegion(_JNIEnv*, _jbyteArray*, int, int, signed char const*)+42)
[art] art/runtime/check_jni.cc:65]   native: #08 pc 00036868   (???)
[art] art/runtime/check_jni.cc:65]   (no managed stack frames)
[art] art/runtime/check_jni.cc:65] 
[art] Suspending all threads took: 18.188ms
[art] Background sticky concurrent mark sweep GC freed 0(0B) AllocSpace objects, 0(0B) LOS objects, 0% free, 511MB/512MB, paused 19.775ms total 51.666ms
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] Clamp target GC heap from 527MB to 512MB
[art] art/runtime/runtime.cc:283] Runtime aborting...
[art] art/runtime/runtime.cc:283] Aborting thread:
[art] art/runtime/runtime.cc:283] "Thread-767" prio=5 tid=22 Native
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12e0f080 self=0xafc36800
[art] art/runtime/runtime.cc:283]   | sysTid=16692 nice=0 cgrp=apps sched=0/0 handle=0x9f90b300
[art] art/runtime/runtime.cc:283]   | state=R schedstat=( 0 0 0 ) utm=94 stm=32 core=0 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0x9e4e1000-0x9e4e3000 stackSize=1020KB
[art] art/runtime/runtime.cc:283]   | held mutexes= "abort lock" "mutator lock"(shared held)
[art] art/runtime/runtime.cc:283]   native: #00 pc 00004c58  /system/lib/libbacktrace_libc++.so (UnwindCurrent::Unwind(unsigned int, ucontext*)+23)
[art] art/runtime/runtime.cc:283]   native: #01 pc 000034c1  /system/lib/libbacktrace_libc++.so (Backtrace::Unwind(unsigned int, ucontext*)+8)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0025918d  /system/lib/libart.so (art::DumpNativeStack(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, int, char const*, art::mirror::ArtMethod*)+84)
[art] art/runtime/runtime.cc:283]   native: #03 pc 0023cd13  /system/lib/libart.so (art::Thread::Dump(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) const+162)
[art] art/runtime/runtime.cc:283]   native: #04 pc 0022c6a9  /system/lib/libart.so (art::AbortState::DumpThread(std::__1::basic_ostream<char, std::__1::char_traits<char> >&, art::Thread*)+32)
[art] art/runtime/runtime.cc:283]   native: #05 pc 0022c94b  /system/lib/libart.so (art::AbortState::Dump(std::__1::basic_ostream<char, std::__1::char_traits<char> >&)+410)
[art] art/runtime/runtime.cc:283]   native: #06 pc 0022cb0b  /system/lib/libart.so (art::Runtime::Abort()+82)
[art] art/runtime/runtime.cc:283]   native: #07 pc 000a7269  /system/lib/libart.so (art::LogMessage::~LogMessage()+1360)
[art] art/runtime/runtime.cc:283]   native: #08 pc 000b1381  /system/lib/libart.so (art::JniAbort(char const*, char const*)+1112)
[art] art/runtime/runtime.cc:283]   native: #09 pc 000b18c5  /system/lib/libart.so (art::JniAbortF(char const*, char const*, ...)+68)
[art] art/runtime/runtime.cc:283]   native: #10 pc 000b4b4d  /system/lib/libart.so (art::ScopedCheck::ScopedCheck(_JNIEnv*, int, char const*)+1324)
[art] art/runtime/runtime.cc:283]   native: #11 pc 000be81b  /system/lib/libart.so (art::CheckJNI::SetByteArrayRegion(_JNIEnv*, _jbyteArray*, int, int, signed char const*)+42)
[art] art/runtime/runtime.cc:283]   native: #12 pc 00036868   (???)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] Pending exception java.lang.OutOfMemoryError thrown by 'unknown throw location'
[art] art/runtime/runtime.cc:283] java.lang.OutOfMemoryError: Failed to allocate a 53302 byte allocation with 19352 free bytes and 18KB until OOM
[art] art/runtime/runtime.cc:283] (Throwable with empty stack trace)Dumping all threads without appropriate locks held: thread list lock mutator lock
[art] art/runtime/runtime.cc:283] All threads:
[art] art/runtime/runtime.cc:283] DALVIK THREADS (28):
[art] art/runtime/runtime.cc:283] "main" prio=5 tid=1 Native
[art] art/runtime/runtime.cc:283]   | group="" sCount=1 dsCount=0 obj=0x732572e0 self=0xb5007800
[art] art/runtime/runtime.cc:283]   | sysTid=16603 nice=0 cgrp=apps sched=0/0 handle=0xb6ff6ec8
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=1343 stm=134 core=1 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xbe56d000-0xbe56f000 stackSize=8MB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 0003a960  /system/lib/libc.so (__epoll_pwait+20)
[art] art/runtime/runtime.cc:283]   native: #01 pc 00014571  /system/lib/libc.so (epoll_pwait+26)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0001457f  /system/lib/libc.so (epoll_wait+6)
[art] art/runtime/runtime.cc:283]   native: #03 pc 000123c3  /system/lib/libutils.so (android::Looper::pollInner(int)+98)
[art] art/runtime/runtime.cc:283]   native: #04 pc 000125ed  /system/lib/libutils.so (android::Looper::pollOnce(int, int*, int*, void**)+92)
[art] art/runtime/runtime.cc:283]   native: #05 pc 0007fb61  /system/lib/libandroid_runtime.so (android::NativeMessageQueue::pollOnce(_JNIEnv*, int)+22)
[art] art/runtime/runtime.cc:283]   native: #06 pc 000b0de7  /data/dalvik-cache/arm/system@framework@boot.oat (Java_android_os_MessageQueue_nativePollOnce__JI+102)
[art] art/runtime/runtime.cc:283]   at android.os.MessageQueue.nativePollOnce(Native method)
[art] art/runtime/runtime.cc:283]   at android.os.MessageQueue.next(MessageQueue.java:143)
[art] art/runtime/runtime.cc:283]   at android.os.Looper.loop(Looper.java:122)
[art] art/runtime/runtime.cc:283]   at android.app.ActivityThread.main(ActivityThread.java:5221)
[art] art/runtime/runtime.cc:283]   at java.lang.reflect.Method.invoke!(Native method)
[art] art/runtime/runtime.cc:283]   at java.lang.reflect.Method.invoke(Method.java:372)
[art] art/runtime/runtime.cc:283]   at com.android.internal.os.ZygoteInit$MethodAndArgsCaller.run(ZygoteInit.java:899)
[art] art/runtime/runtime.cc:283]   at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:694)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "Heap thread pool worker thread 2" prio=5 tid=2 Native (still starting up)
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x0 self=0xaff0e400
[art] art/runtime/runtime.cc:283]   | sysTid=16611 nice=0 cgrp=apps sched=0/0 handle=0xac4a4b00
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=1 stm=0 core=2 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb4634000-0xb4636000 stackSize=1020KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e98  /system/lib/libc.so (syscall+28)
[art] art/runtime/runtime.cc:283]   native: #01 pc 000a9fa3  /system/lib/libart.so (art::ConditionVariable::Wait(art::Thread*)+98)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0024714d  /system/lib/libart.so (art::ThreadPool::GetTask(art::Thread*)+64)
[art] art/runtime/runtime.cc:283]   native: #03 pc 002470e7  /system/lib/libart.so (art::ThreadPoolWorker::Run()+62)
[art] art/runtime/runtime.cc:283]   native: #04 pc 002479d5  /system/lib/libart.so (art::ThreadPoolWorker::Callback(void*)+60)
[art] art/runtime/runtime.cc:283]   native: #05 pc 000162e3  /system/lib/libc.so (__pthread_start(void*)+30)
[art] art/runtime/runtime.cc:283]   native: #06 pc 000142d3  /system/lib/libc.so (__start_thread+6)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "Heap thread pool worker thread 1" prio=5 tid=3 Native (still starting up)
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x0 self=0xafc0e400
[art] art/runtime/runtime.cc:283]   | sysTid=16610 nice=0 cgrp=apps sched=0/0 handle=0xac4a4880
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=0 stm=1 core=2 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb4734000-0xb4736000 stackSize=1020KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e98  /system/lib/libc.so (syscall+28)
[art] art/runtime/runtime.cc:283]   native: #01 pc 000a9fa3  /system/lib/libart.so (art::ConditionVariable::Wait(art::Thread*)+98)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0024714d  /system/lib/libart.so (art::ThreadPool::GetTask(art::Thread*)+64)
[art] art/runtime/runtime.cc:283]   native: #03 pc 002470e7  /system/lib/libart.so (art::ThreadPoolWorker::Run()+62)
[art] art/runtime/runtime.cc:283]   native: #04 pc 002479d5  /system/lib/libart.so (art::ThreadPoolWorker::Callback(void*)+60)
[art] art/runtime/runtime.cc:283]   native: #05 pc 000162e3  /system/lib/libc.so (__pthread_start(void*)+30)
[art] art/runtime/runtime.cc:283]   native: #06 pc 000142d3  /system/lib/libc.so (__start_thread+6)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "Heap thread pool worker thread 0" prio=5 tid=4 Native (still starting up)
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x0 self=0xaf90e400
[art] art/runtime/runtime.cc:283]   | sysTid=16609 nice=0 cgrp=apps sched=0/0 handle=0xac4a4600
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=0 stm=1 core=2 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb4834000-0xb4836000 stackSize=1020KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e98  /system/lib/libc.so (syscall+28)
[art] art/runtime/runtime.cc:283]   native: #01 pc 000a9fa3  /system/lib/libart.so (art::ConditionVariable::Wait(art::Thread*)+98)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0024714d  /system/lib/libart.so (art::ThreadPool::GetTask(art::Thread*)+64)
[art] art/runtime/runtime.cc:283]   native: #03 pc 002470e7  /system/lib/libart.so (art::ThreadPoolWorker::Run()+62)
[art] art/runtime/runtime.cc:283]   native: #04 pc 002479d5  /system/lib/libart.so (art::ThreadPoolWorker::Callback(void*)+60)
[art] art/runtime/runtime.cc:283]   native: #05 pc 000162e3  /system/lib/libc.so (__pthread_start(void*)+30)
[art] art/runtime/runtime.cc:283]   native: #06 pc 000142d3  /system/lib/libc.so (__start_thread+6)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "Signal Catcher" prio=5 tid=5 WaitingInMainSignalCatcherLoop
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12c00080 self=0xaf60e400
[art] art/runtime/runtime.cc:283]   | sysTid=16612 nice=0 cgrp=apps sched=0/0 handle=0xac4a4d80
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=0 stm=0 core=1 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb4524000-0xb4526000 stackSize=1012KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 0003a0b4  /system/lib/libc.so (__rt_sigtimedwait+12)
[art] art/runtime/runtime.cc:283]   native: #01 pc 00017a25  /system/lib/libc.so (sigwait+24)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0022fe6b  /system/lib/libart.so (art::SignalCatcher::WaitForSignal(art::Thread*, art::SignalSet&)+86)
[art] art/runtime/runtime.cc:283]   native: #03 pc 002319f5  /system/lib/libart.so (art::SignalCatcher::Run(void*)+212)
[art] art/runtime/runtime.cc:283]   native: #04 pc 000162e3  /system/lib/libc.so (__pthread_start(void*)+30)
[art] art/runtime/runtime.cc:283]   native: #05 pc 000142d3  /system/lib/libc.so (__start_thread+6)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "JDWP" prio=5 tid=6 WaitingInMainDebuggerLoop
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12c05080 self=0xaf30e400
[art] art/runtime/runtime.cc:283]   | sysTid=16613 nice=0 cgrp=apps sched=0/0 handle=0xac4a5000
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=0 stm=0 core=0 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb4420000-0xb4422000 stackSize=1012KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00039a20  /system/lib/libc.so (__pselect6+20)
[art] art/runtime/runtime.cc:283]   native: #01 pc 00015abb  /system/lib/libc.so (select+60)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0029d86f  /system/lib/libart.so (art::JDWP::JdwpAdbState::ProcessIncoming()+202)
[art] art/runtime/runtime.cc:283]   native: #03 pc 00195fe5  /system/lib/libart.so (art::JDWP::JdwpState::Run()+356)
[art] art/runtime/runtime.cc:283]   native: #04 pc 00197531  /system/lib/libart.so (art::JDWP::StartJdwpThread(void*)+16)
[art] art/runtime/runtime.cc:283]   native: #05 pc 000162e3  /system/lib/libc.so (__pthread_start(void*)+30)
[art] art/runtime/runtime.cc:283]   native: #06 pc 000142d3  /system/lib/libc.so (__start_thread+6)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "Binder_1" prio=5 tid=7 Native
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12c59080 self=0xb3c06400
[art] art/runtime/runtime.cc:283]   | sysTid=16619 nice=0 cgrp=apps sched=0/0 handle=0xac4a5f00
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=9 stm=10 core=1 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb3d0a000-0xb3d0c000 stackSize=1012KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00039984  /system/lib/libc.so (__ioctl+8)
[art] art/runtime/runtime.cc:283]   native: #01 pc 0005177d  /system/lib/libc.so (ioctl+14)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0001f39b  /system/lib/libbinder.so (android::IPCThreadState::talkWithDriver(bool)+138)
[art] art/runtime/runtime.cc:283]   native: #03 pc 0001f88b  /system/lib/libbinder.so (android::IPCThreadState::getAndExecuteCommand()+6)
[art] art/runtime/runtime.cc:283]   native: #04 pc 0001f8ed  /system/lib/libbinder.so (android::IPCThreadState::joinThreadPool(bool)+48)
[art] art/runtime/runtime.cc:283]   native: #05 pc 00023a5b  /system/lib/libbinder.so (???)
[art] art/runtime/runtime.cc:283]   native: #06 pc 000104d5  /system/lib/libutils.so (android::Thread::_threadLoop(void*)+112)
[art] art/runtime/runtime.cc:283]   native: #07 pc 0005df4d  /system/lib/libandroid_runtime.so (android::AndroidRuntime::javaThreadShell(void*)+72)
[art] art/runtime/runtime.cc:283]   native: #08 pc 00010045  /system/lib/libutils.so (???)
[art] art/runtime/runtime.cc:283]   native: #09 pc 000162e3  /system/lib/libc.so (__pthread_start(void*)+30)
[art] art/runtime/runtime.cc:283]   native: #10 pc 000142d3  /system/lib/libc.so (__start_thread+6)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "GCDaemon" prio=5 tid=8 WaitingPerformingGc
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12c27200 self=0xac4b5400
[art] art/runtime/runtime.cc:283]   | sysTid=16618 nice=0 cgrp=apps sched=0/0 handle=0xac4a5c80
[art] art/runtime/runtime.cc:283]   | state=R schedstat=( 0 0 0 ) utm=62 stm=9 core=1 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb3f06000-0xb3f08000 stackSize=1036KB
[art] art/runtime/runtime.cc:283]   | held mutexes= "heap bitmap lock"(exclusive held) "mutator lock"(shared held)
[art] art/runtime/runtime.cc:283]   native: #00 pc 00129f78  /system/lib/libart.so (void art::mirror::Object::VisitFieldsReferences<false, false, art::gc::collector::MarkObjectVisitor>(unsigned int, art::gc::collector::MarkObjectVisitor const&)+43)
[art] art/runtime/runtime.cc:283]   native: #01 pc 0012a0e9  /system/lib/libart.so (void art::mirror::Object::VisitReferences<false, (art::VerifyObjectFlags)0, art::gc::collector::MarkObjectVisitor, art::gc::collector::DelayReferenceReferentVisitor>(art::gc::collector::MarkObjectVisitor const&, art::gc::collector::DelayReferenceReferentVisitor const&)+180)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0012a3b1  /system/lib/libart.so (art::gc::collector::MarkSweep::ProcessMarkStack(bool)+160)
[art] art/runtime/runtime.cc:283]   native: #03 pc 0012a46d  /system/lib/libart.so (art::gc::collector::MarkSweep::MarkReachableObjects()+40)
[art] art/runtime/runtime.cc:283]   native: #04 pc 0012b39f  /system/lib/libart.so (art::gc::collector::MarkSweep::MarkingPhase()+122)
[art] art/runtime/runtime.cc:283]   native: #05 pc 0012b4fb  /system/lib/libart.so (art::gc::collector::MarkSweep::RunPhases()+290)
[art] art/runtime/runtime.cc:283]   native: #06 pc 00122187  /system/lib/libart.so (art::gc::collector::GarbageCollector::Run(art::gc::GcCause, bool)+246)
[art] art/runtime/runtime.cc:283]   native: #07 pc 001421df  /system/lib/libart.so (art::gc::Heap::CollectGarbageInternal(art::gc::collector::GcType, art::gc::GcCause, bool)+1438)
[art] art/runtime/runtime.cc:283]   native: #08 pc 001435f1  /system/lib/libart.so (art::gc::Heap::ConcurrentGC(art::Thread*)+56)
[art] art/runtime/runtime.cc:283]   native: #09 pc 000003d3  /data/dalvik-cache/arm/system@framework@boot.oat (Java_dalvik_system_VMRuntime_concurrentGC__+82)
[art] art/runtime/runtime.cc:283]   at dalvik.system.VMRuntime.concurrentGC(Native method)
[art] art/runtime/runtime.cc:283]   at java.lang.Daemons$GCDaemon.run(Daemons.java:345)
[art] art/runtime/runtime.cc:283]   at java.lang.Thread.run(Thread.java:818)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "HeapTrimmerDaemon" prio=5 tid=9 Sleeping
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12c271a0 self=0xac4b5000
[art] art/runtime/runtime.cc:283]   | sysTid=16617 nice=0 cgrp=apps sched=0/0 handle=0xac4a5a00
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=0 stm=0 core=2 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb400a000-0xb400c000 stackSize=1036KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00039b08  /system/lib/libc.so (nanosleep+12)
[art] art/runtime/runtime.cc:283]   native: #01 pc 00024659  /system/lib/libc.so (usleep+32)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0014010b  /system/lib/libart.so (art::gc::Heap::DoPendingTransitionOrTrim()+690)
[art] art/runtime/runtime.cc:283]   native: #03 pc 000003d3  /data/dalvik-cache/arm/system@framework@boot.oat (Java_dalvik_system_VMRuntime_trimHeap__+82)
[art] art/runtime/runtime.cc:283]   at dalvik.system.VMRuntime.trimHeap(Native method)
[art] art/runtime/runtime.cc:283]   - sleeping on an unknown object
[art] art/runtime/runtime.cc:283]   at java.lang.Daemons$HeapTrimmerDaemon.run(Daemons.java:312)
[art] art/runtime/runtime.cc:283]   at java.lang.Thread.run(Thread.java:818)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "FinalizerWatchdogDaemon" prio=5 tid=10 Sleeping
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12c27140 self=0xac4b4c00
[art] art/runtime/runtime.cc:283]   | sysTid=16616 nice=0 cgrp=apps sched=0/0 handle=0xac4a5780
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=0 stm=0 core=2 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb410e000-0xb4110000 stackSize=1036KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e9c  /system/lib/libc.so (syscall+32)
[art] art/runtime/runtime.cc:283]   native: #01 pc 000a9d2f  /system/lib/libart.so (art::ConditionVariable::TimedWait(art::Thread*, long long, int)+118)
[art] art/runtime/runtime.cc:283]   native: #02 pc 001fbafd  /system/lib/libart.so (art::Monitor::Wait(art::Thread*, long long, int, bool, art::ThreadState)+504)
[art] art/runtime/runtime.cc:283]   native: #03 pc 001fd3d5  /system/lib/libart.so (art::Monitor::Wait(art::Thread*, art::mirror::Object*, long long, int, bool, art::ThreadState)+136)
[art] art/runtime/runtime.cc:283]   native: #04 pc 0007145d  /data/dalvik-cache/arm/system@framework@boot.oat (Java_java_lang_Thread_sleep__Ljava_lang_Object_2JI+132)
[art] art/runtime/runtime.cc:283]   at java.lang.Thread.sleep!(Native method)
[art] art/runtime/runtime.cc:283]   - sleeping on <0x3e0d6aff> (a java.lang.Object)
[art] art/runtime/runtime.cc:283]   at java.lang.Thread.sleep(Thread.java:1031)
[art] art/runtime/runtime.cc:283]   - locked <0x3e0d6aff> (a java.lang.Object)
[art] art/runtime/runtime.cc:283]   at java.lang.Thread.sleep(Thread.java:985)
[art] art/runtime/runtime.cc:283]   at java.lang.Daemons$FinalizerWatchdogDaemon.sleepFor(Daemons.java:257)
[art] art/runtime/runtime.cc:283]   at java.lang.Daemons$FinalizerWatchdogDaemon.waitForFinalization(Daemons.java:268)
[art] art/runtime/runtime.cc:283]   at java.lang.Daemons$FinalizerWatchdogDaemon.run(Daemons.java:216)
[art] art/runtime/runtime.cc:283]   at java.lang.Thread.run(Thread.java:818)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "FinalizerDaemon" prio=5 tid=11 Waiting
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12c270e0 self=0xac4b4800
[art] art/runtime/runtime.cc:283]   | sysTid=16615 nice=0 cgrp=apps sched=0/0 handle=0xac4a5500
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=6 stm=3 core=1 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb4212000-0xb4214000 stackSize=1036KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e98  /system/lib/libc.so (syscall+28)
[art] art/runtime/runtime.cc:283]   native: #01 pc 000a9fa3  /system/lib/libart.so (art::ConditionVariable::Wait(art::Thread*)+98)
[art] art/runtime/runtime.cc:283]   native: #02 pc 001fbd11  /system/lib/libart.so (art::Monitor::Wait(art::Thread*, long long, int, bool, art::ThreadState)+1036)
[art] art/runtime/runtime.cc:283]   native: #03 pc 001fd3d5  /system/lib/libart.so (art::Monitor::Wait(art::Thread*, art::mirror::Object*, long long, int, bool, art::ThreadState)+136)
[art] art/runtime/runtime.cc:283]   native: #04 pc 0020ae09  /system/lib/libart.so (art::Object_waitJI(_JNIEnv*, _jobject*, long long, int)+44)
[art] art/runtime/runtime.cc:283]   native: #05 pc 000005f7  /data/dalvik-cache/arm/system@framework@boot.oat (Java_java_lang_Object_wait__JI+102)
[art] art/runtime/runtime.cc:283]   at java.lang.Object.wait!(Native method)
[art] art/runtime/runtime.cc:283]   - waiting on <0x028a70cc> (a java.lang.ref.ReferenceQueue)
[art] art/runtime/runtime.cc:283]   at java.lang.Object.wait(Object.java:422)
[art] art/runtime/runtime.cc:283]   at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:101)
[art] art/runtime/runtime.cc:283]   - locked <0x028a70cc> (a java.lang.ref.ReferenceQueue)
[art] art/runtime/runtime.cc:283]   at java.lang.ref.ReferenceQueue.remove(ReferenceQueue.java:72)
[art] art/runtime/runtime.cc:283]   at java.lang.Daemons$FinalizerDaemon.run(Daemons.java:173)
[art] art/runtime/runtime.cc:283]   at java.lang.Thread.run(Thread.java:818)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "Binder_2" prio=5 tid=12 Native
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12c73080 self=0xa4406400
[art] art/runtime/runtime.cc:283]   | sysTid=16620 nice=0 cgrp=apps sched=0/0 handle=0xb3c26080
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=10 stm=9 core=1 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb3b04000-0xb3b06000 stackSize=1012KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00039984  /system/lib/libc.so (__ioctl+8)
[art] art/runtime/runtime.cc:283]   native: #01 pc 0005177d  /system/lib/libc.so (ioctl+14)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0001f39b  /system/lib/libbinder.so (android::IPCThreadState::talkWithDriver(bool)+138)
[art] art/runtime/runtime.cc:283]   native: #03 pc 0001f88b  /system/lib/libbinder.so (android::IPCThreadState::getAndExecuteCommand()+6)
[art] art/runtime/runtime.cc:283]   native: #04 pc 0001f8ed  /system/lib/libbinder.so (android::IPCThreadState::joinThreadPool(bool)+48)
[art] art/runtime/runtime.cc:283]   native: #05 pc 00023a5b  /system/lib/libbinder.so (???)
[art] art/runtime/runtime.cc:283]   native: #06 pc 000104d5  /system/lib/libutils.so (android::Thread::_threadLoop(void*)+112)
[art] art/runtime/runtime.cc:283]   native: #07 pc 0005df4d  /system/lib/libandroid_runtime.so (android::AndroidRuntime::javaThreadShell(void*)+72)
[art] art/runtime/runtime.cc:283]   native: #08 pc 00010045  /system/lib/libutils.so (???)
[art] art/runtime/runtime.cc:283]   native: #09 pc 000162e3  /system/lib/libc.so (__pthread_start(void*)+30)
[art] art/runtime/runtime.cc:283]   native: #10 pc 000142d3  /system/lib/libc.so (__start_thread+6)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "ReferenceQueueDaemon" prio=5 tid=13 Waiting
[art] art/runtime/runtime.cc:283]   | group="" sCount=1 dsCount=0 obj=0x12c27080 self=0xac4b4400
[art] art/runtime/runtime.cc:283]   | sysTid=16614 nice=0 cgrp=apps sched=0/0 handle=0xac4a5280
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=1 stm=3 core=0 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xb4316000-0xb4318000 stackSize=1036KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e98  /system/lib/libc.so (syscall+28)
[art] art/runtime/runtime.cc:283]   native: #01 pc 000a9fa3  /system/lib/libart.so (art::ConditionVariable::Wait(art::Thread*)+98)
[art] art/runtime/runtime.cc:283]   native: #02 pc 001fbd11  /system/lib/libart.so (art::Monitor::Wait(art::Thread*, long long, int, bool, art::ThreadState)+1036)
[art] art/runtime/runtime.cc:283]   native: #03 pc 001fd3d5  /system/lib/libart.so (art::Monitor::Wait(art::Thread*, art::mirror::Object*, long long, int, bool, art::ThreadState)+136)
[art] art/runtime/runtime.cc:283]   native: #04 pc 0020ae41  /system/lib/libart.so (art::Object_wait(_JNIEnv*, _jobject*)+40)
[art] art/runtime/runtime.cc:283]   native: #05 pc 000003d3  /data/dalvik-cache/arm/system@framework@boot.oat (Java_java_lang_Object_wait__+82)
[art] art/runtime/runtime.cc:283]   at java.lang.Object.wait!(Native method)
[art] art/runtime/runtime.cc:283]   - waiting on <0x0d185f15> (a java.lang.Class<java.lang.ref.ReferenceQueue>)
[art] art/runtime/runtime.cc:283]   at java.lang.Daemons$ReferenceQueueDaemon.run(Daemons.java:133)
[art] art/runtime/runtime.cc:283]   - locked <0x0d185f15> (a java.lang.Class<java.lang.ref.ReferenceQueue>)
[art] art/runtime/runtime.cc:283]   at java.lang.Thread.run(Thread.java:818)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "Thread-757" prio=5 tid=14 Native
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12d01080 self=0xa0112800
[art] art/runtime/runtime.cc:283]   | sysTid=16634 nice=0 cgrp=apps sched=0/0 handle=0xac4a6400
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=2 stm=0 core=0 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xa027e000-0xa0280000 stackSize=1020KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e98  /system/lib/libc.so (syscall+28)
[art] art/runtime/runtime.cc:283]   native: #01 pc 00016b4f  /system/lib/libc.so (__futex_wait_ex+42)
[art] art/runtime/runtime.cc:283]   native: #02 pc 00051be1  /system/lib/libc.so (sem_wait+50)
[art] art/runtime/runtime.cc:283]   native: #03 pc 001e81d0  /data/app/be.x.x-1/lib/arm/libmonosgen-2.0.so (mono_sem_wait+16)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "Thread-759" prio=5 tid=15 Native
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12d3d080 self=0x9f913800
[art] art/runtime/runtime.cc:283]   | sysTid=16644 nice=0 cgrp=apps sched=0/0 handle=0xac4a2d00
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=1 stm=0 core=1 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0xa1012000-0xa1014000 stackSize=124KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00039b08  /system/lib/libc.so (nanosleep+12)
[art] art/runtime/runtime.cc:283]   native: #01 pc 001db388  /data/app/be.x.x-1/lib/arm/libmonosgen-2.0.so (???)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "Thread-761" prio=5 tid=17 Native
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12da1080 self=0x9fe7b400
[art] art/runtime/runtime.cc:283]   | sysTid=16652 nice=0 cgrp=apps sched=0/0 handle=0xac4a2580
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=3 stm=2 core=0 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0x9f702000-0x9f704000 stackSize=1020KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e9c  /system/lib/libc.so (syscall+32)
[art] art/runtime/runtime.cc:283]   native: #01 pc 000161b5  /system/lib/libc.so (__pthread_cond_timedwait_relative(pthread_cond_t*, pthread_mutex_t*, timespec const*)+56)
[art] art/runtime/runtime.cc:283]   native: #02 pc 000161f9  /system/lib/libc.so (__pthread_cond_timedwait(pthread_cond_t*, pthread_mutex_t*, timespec const*, int)+28)
[art] art/runtime/runtime.cc:283]   native: #03 pc 001c8e98  /data/app/be.x.x-1/lib/arm/libmonosgen-2.0.so (???)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283] "RenderThread" prio=5 tid=18 Native
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12db5080 self=0xaff2a400
[art] art/runtime/runtime.cc:283]   | sysTid=16653 nice=-4 cgrp=apps sched=0/0 handle=0xac4a2f80
[art] art/runtime/runtime.cc:283]   | state=R schedstat=( 0 0 0 ) utm=1120 stm=277 core=1 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0x9f304000-0x9f306000 stackSize=1012KB
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 0003a960  /system/lib/libc.so (__epoll_pwait+20)
[art] art/runtime/runtime.cc:283]   native: #01 pc 00014571  /system/lib/libc.so (epoll_pwait+26)
[art] art/runtime/runtime.cc:283]   native: #02 pc 0001457f  /system/lib/libc.so (epoll_wait+6)
[art] art/runtime/runtime.cc:283]   native: #03 pc 000123c3  /system/lib/libutils.so (android::Looper::pollInner(int)+98)
[art] art/runtime/runtime.cc:283]   native: #04 pc 000125ed  /system/lib/libutils.so (android::Looper::pollOnce(int, int*, int*, void**)+92)
[art] art/runtime/runtime.cc:283]   native: #05 pc 0003e265  /system/lib/libhwui.so (android::uirenderer::renderthread::RenderThread::threadLoop()+44)
[art] art/runtime/runtime.cc:283]   native: #06 pc 000104d5  /system/lib/libutils.so (android::Thread::_threadLoop(void*)+112)
[art] art/runtime/runtime.cc:283]   native: #07 pc 0005df4d  /system/lib/libandroid_runtime.so (android::AndroidRuntime::javaThreadShell(void*)+72)
[art] art/runtime/runtime.cc:283]   native: #07 pc 000142d3  /system/lib/libc.so (__start_thread+6)
[art] art/runtime/runtime.cc:283] "hwuiTask2" prio=5 tid=21 Native
[art] art/runtime/runtime.cc:283]   | sysTid=16669 nice=-2 cgrp=apps sched=0/0 handle=0x9e6bad00
[art] art/runtime/runtime.cc:283]   | stack=0x9cbf5000-0x9cbf7000 stackSize=1012KB
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e98  /system/lib/libc.so (syscall+28)
[art] art/runtime/runtime.cc:283]   native: #02 pc 00014d6d  /system/lib/libhwui.so (???)
[art] art/runtime/runtime.cc:283]   native: #04 pc 0005df4d  /system/lib/libandroid_runtime.so (android::AndroidRuntime::javaThreadShell(void*)+72)
[art] art/runtime/runtime.cc:283]   native: #06 pc 000162e3  /system/lib/libc.so (__pthread_start(void*)+30)
[art] art/runtime/runtime.cc:283] "Thread-767" prio=5 tid=22 Native
[art] art/runtime/runtime.cc:283]   | sysTid=16692 nice=0 cgrp=apps sched=0/0 handle=0x9f90b300
[art] art/runtime/runtime.cc:283]   | stack=0x9e4e1000-0x9e4e3000 stackSize=1020KB
[art] art/runtime/runtime.cc:283]   native: #00 pc 00004c58  /system/lib/libbacktrace_libc++.so (UnwindCurrent::Unwind(unsigned int, ucontext*)+23)
[art] art/runtime/runtime.cc:283]   native: #01 pc 000034c1  /system/lib/libbacktrace_libc++.so (Backtrace::Unwind(unsigned int, ucontext*)+8)
[art] art/runtime/runtime.cc:283]   native: #03 pc 0023cd13  /system/lib/libart.so (art::Thread::Dump(std::__1::basic_ostream<char, std::__1::char_traits<char> >&) const+162)
[art] art/runtime/runtime.cc:283]   native: #05 pc 0022c8c1  /system/lib/libart.so (art::AbortState::Dump(std::__1::basic_ostream<char, std::__1::char_traits<char> >&)+272)
[art] art/runtime/runtime.cc:283]   native: #06 pc 0022cb0b  /system/lib/libart.so (art::Runtime::Abort()+82)
[art] art/runtime/runtime.cc:283]   native: #08 pc 000b1381  /system/lib/libart.so (art::JniAbort(char const*, char const*)+1112)
[art] art/runtime/runtime.cc:283]   native: #10 pc 000b4b4d  /system/lib/libart.so (art::ScopedCheck::ScopedCheck(_JNIEnv*, int, char const*)+1324)
[art] art/runtime/runtime.cc:283]   native: #12 pc 00036868   (???)
[art] art/runtime/runtime.cc:283] 
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12e14080 self=0xafe1e800
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=137 stm=37 core=1 HZ=100
[art] art/runtime/runtime.cc:283]   | stack=0x9ddf2000-0x9ddf4000 stackSize=1020KB
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e9c  /system/lib/libc.so (syscall+32)
[art] art/runtime/runtime.cc:283]   native: #02 pc 000161f9  /system/lib/libc.so (__pthread_cond_timedwait(pthread_cond_t*, pthread_mutex_t*, timespec const*, int)+28)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] "Thread-769" prio=5 tid=24 Native
[art] art/runtime/runtime.cc:283]   | sysTid=16705 nice=0 cgrp=apps sched=0/0 handle=0x9f90b800
[art] art/runtime/runtime.cc:283]   | stack=0x9d9f2000-0x9d9f4000 stackSize=1020KB
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e9c  /system/lib/libc.so (syscall+32)
[art] art/runtime/runtime.cc:283]   native: #01 pc 00016b4f  /system/lib/libc.so (__futex_wait_ex+42)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] "disconnect check" prio=5 tid=25 TimedWaiting
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12dd0f20 self=0x9f613000
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=0 stm=0 core=0 HZ=100
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #01 pc 000a9d2f  /system/lib/libart.so (art::ConditionVariable::TimedWait(art::Thread*, long long, int)+118)
[art] art/runtime/runtime.cc:283]   native: #03 pc 001fd3d5  /system/lib/libart.so (art::Monitor::Wait(art::Thread*, art::mirror::Object*, long long, int, bool, art::ThreadState)+136)
[art] art/runtime/runtime.cc:283]   native: #05 pc 000005f7  /data/dalvik-cache/arm/system@framework@boot.oat (Java_java_lang_Object_wait__JI+102)
[art] art/runtime/runtime.cc:283]   at java.lang.Object.wait!(Native method)
[art] art/runtime/runtime.cc:283]   at java.lang.Object.wait(Object.java:422)
[art] art/runtime/runtime.cc:283]   - locked <0x3157b81b> (a java.util.Timer$TimerImpl)
[art] art/runtime/runtime.cc:283] "Thread-772" prio=5 tid=26 Native
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12e57080 self=0xaf51e800
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=0 stm=6 core=0 HZ=100
[art] art/runtime/runtime.cc:283]   native: #02 pc 0001457f  /system/lib/libc.so (epoll_wait+6)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] "Thread-773" prio=5 tid=27 Native
[art] art/runtime/runtime.cc:283]   | sysTid=16757 nice=0 cgrp=apps sched=0/0 handle=0x9fa07a80
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=3 stm=11 core=0 HZ=100
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #01 pc 00016b4f  /system/lib/libc.so (__futex_wait_ex+42)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] "OkHttp ConnectionPool" prio=5 tid=28 TimedWaiting
[art] art/runtime/runtime.cc:283]   | sysTid=16869 nice=0 cgrp=apps sched=0/0 handle=0xafc2ba80
[art] art/runtime/runtime.cc:283]   | stack=0x9bbfe000-0x9bc00000 stackSize=1036KB
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e9c  /system/lib/libc.so (syscall+32)
[art] art/runtime/runtime.cc:283]   native: #02 pc 001fbafd  /system/lib/libart.so (art::Monitor::Wait(art::Thread*, long long, int, bool, art::ThreadState)+504)
[art] art/runtime/runtime.cc:283]   native: #04 pc 0020ae09  /system/lib/libart.so (art::Object_waitJI(_JNIEnv*, _jobject*, long long, int)+44)
[art] art/runtime/runtime.cc:283]   at java.lang.Object.wait!(Native method)
[art] art/runtime/runtime.cc:283]   at java.lang.Thread.parkFor(Thread.java:1220)
[art] art/runtime/runtime.cc:283]   - locked <0x3c53cdb8> (a java.lang.Object)
[art] art/runtime/runtime.cc:283]   at java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:197)
[art] art/runtime/runtime.cc:283]   at java.util.concurrent.LinkedBlockingQueue.poll(LinkedBlockingQueue.java:435)
[art] art/runtime/runtime.cc:283]   at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1097)
[art] art/runtime/runtime.cc:283]   | group="" sCount=0 dsCount=0 obj=0x12c73260 self=0xa01b0800
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=64 stm=25 core=2 HZ=100
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #00 pc 00012e9c  /system/lib/libc.so (syscall+32)
[art] art/runtime/runtime.cc:283]   native: #01 pc 00016b4f  /system/lib/libc.so (__futex_wait_ex+42)
[art] art/runtime/runtime.cc:283]   native: #03 pc 001e82c0  /data/app/be.x.x-1/lib/arm/libmonosgen-2.0.so (mono_sem_timedwait+176)
[art] art/runtime/runtime.cc:283]   (no managed stack frames)
[art] art/runtime/runtime.cc:283] "Binder_3" prio=5 tid=16 Native
[art] art/runtime/runtime.cc:283]   | state=S schedstat=( 0 0 0 ) utm=0 stm=0 core=0 HZ=100
[art] art/runtime/runtime.cc:283]   | held mutexes=
[art] art/runtime/runtime.cc:283]   native: #01 pc 0005177d  /system/lib/libc.so (ioctl+14)
[libc] Fatal signal 6 (SIGABRT), code -6 in tid 16692 (Thread-767)
```

**Comment by sescandell** [comment]
Some updates about the Volley implementation on KitKat device: it's now working well... without changing anything... I'm so going to check how to integrate Volley principle into MvvmCross through the DownloadPlugin or by creating a project mapping Volley method to the Xamarin world.

**Comment by slodge** [comment]
Stephane

Sorry for the "radio silence" from me on this.... my excuse is I was in France skiing when this came in :)

One suggestion that's just come through on https://jabbr.net/#/rooms/mvvmcross is that if this is down to the Http layer, then maybe ModernHttpClient might help here - as *maybe* that already uses the Volley library.

Regardless, you're doing a **fab** job - so please do push forwards with your investigations and coding - **love it** :+1:

**Comment by sescandell** [comment]
Hi there,

Sorry for the "long silence".
I've found a workaround for this issue using an external library.
You could find everything here: https://gist.github.com/sescandell/e2079e2444a50dc3b23e

Basically, I created my own ImageVien Android component using the component AQueryXamarin to manage requests and cache.

This is working very well on both platform 4.4 and 5.X+

I'm not sure this could be integrated into MvvmCross because I'm using here an external library. But I think this confirm our last discussion. Maybe we should get some inspiration from this project (and Volley and OkHttp).

**Comment by azchohfi** [comment]
Any update on this? I'm having the same problem here. And by the way, I think this would be a good time to implement the shouldCache param:
```
 public MvxImage<Bitmap> Load(string localPath, bool shouldCache /* ignored here */)
```

**Comment by SeeD-Seifer** [comment]
Just for the record, implementation of ```shouldCache``` parameter: https://github.com/MvvmCross/MvvmCross/pull/947

**Comment by bjarkeorama** [comment]
Hi Guys, I am currently hitting this issue as well. Any updates on this or what to do? I tried sescandell workaround but it still dies on my nexus with android 5.1

**Comment by sescandell** [comment]
Hi @bjarkeorama 

Do ou have any sample on how do you implemented it?

I'll check my gist is still working on 5.1 and let you know.

**Comment by cwjtojava** [comment]
Crash in the Android Lollipop.
Details:
06-19 12:50:35.500: E/art(1224): Throwing OutOfMemoryError "Failed to allocate a 262156 byte allocation with 35652 free bytes and 34KB until OOM"

**Comment by CSkoubo** [comment]
What have you guys set your MaxInMemoryBytes to? i increased mine to use 50mb instead of the default 4mb, also enabled the DisposeOnRemoveFromCache This helped ALOT on my Nexus 6, and i don't see the error any more. But our app is still using alot of memory.

**Comment by SeeD-Seifer** [comment]
Finally got time to investigate the issue.

The reason of memory leak is: **Mono does not free bitmap references so that Java can't free memory**

I've found the working solution in https://github.com/molinch/FFImageLoading

Basically the idea is to wrap with try-catch BitmapFactory methods and force Mono garbage collection in case there is Java.Lang.OutOfMemoryError thrown.
I've tested solution with Memory Profiler and it works.

Fixing DownloadCache plugin will reduce OOM chances. Though there is should be a more generic way to catch Java's OOM and force Mono GC (which I have not found yet).
**OnTrimMemory** should be a good place to force Mono GC, but for some reason it is not called when OOM is about to be thrown.

Going to prepare a PR with a fix for DownloadCache, but this won't fix OOM completely.

**Comment by js8794** [comment]
@SeeD-Seifer, I'm experiencing this issue with my app and was wondering the state of your PR? I didn't see it when I took a quick glance. @slodge, wondering if you have any more insight on this?

**Comment by js8794** [comment]
I tried uping my MaxInMemoryBytes as @CSkoubo suggested to no avail (Didn't really expect it to work). Also started looking at @SeeD-Seifer's change. Not sure that it sounds like the best solution, but anything is better than crashing. Looks like the Plugins repo has had a lot of churn lately.. I was hoping to be able to build the DownloadCache locally with his change and throw it in my app in place of the dll I got from NuGet, however I ran into a lot of dependency/capability issues. Anyhow, any help would be a appreciated. (If you couldn't tell, I am not the most experienced in the MvvmCross world)

**Comment by PelleRavn** [comment]
Any news in this? 

I'm current experience the exact same issue on newer Android versions, and came to the same conclusion as @SeeD-Seifer with the BitmapFactory leaking.

**Comment by UKDeveloper99** [comment]
Same as above.
